// @thirdweb-dev/wallets/evm/wallets/device-wallet/dist/thirdweb-dev-wallets-evm-wallets-device-wallet.esm.js
import { _ as _classPrivateFieldInitSpec, a as _classPrivateFieldSet, b as _classPrivateFieldGet } from '../../../../dist/classPrivateFieldSet-0ee02dfd.esm.js';
import { _ as _defineProperty } from '../../../../dist/defineProperty-e24c82ea.esm.js';
import { AbstractWallet } from '../../abstract/dist/thirdweb-dev-wallets-evm-wallets-abstract.esm.js';
import { a as AbstractClientWallet, c as createAsyncLocalStorage } from '../../../../dist/base-9746420d.esm.js';
import { Ethereum, defaultChains } from '@thirdweb-dev/chains';
import { ethers } from 'ethers';
import '../../../../dist/checkPrivateRedeclaration-a6ec2e61.esm.js';
import 'eventemitter3';
import '../../../../dist/classPrivateMethodGet-fb5087d9.esm.js';

// no need for prefixing here - AsyncStorage is already namespaced
const STORAGE_KEY_DATA = "data";
const STORAGE_KEY_ADDR = "address";
var _walletImpl = /*#__PURE__*/new WeakMap();
class DeviceBrowserWallet extends AbstractClientWallet {
  get walletName() {
    return "Device Wallet";
  }
  constructor(options) {
    super(DeviceBrowserWallet.id, {
      ...options
    });
    _defineProperty(this, "connector", void 0);
    _classPrivateFieldInitSpec(this, _walletImpl, {
      writable: true,
      value: void 0
    });
    _defineProperty(this, "options", void 0);
    this.options = options;
  }
  async getConnector() {
    if (!this.connector) {
      // import the connector dynamically
      const {
        DeviceWalletConnector
      } = await import('../../../connectors/device-wallet/dist/thirdweb-dev-wallets-evm-connectors-device-wallet.esm.js');
      let wallet;
      switch (this.options.storageType) {
        case "asyncStore":
          wallet = await DeviceWalletImpl.fromAsyncStorage(this.options.storage || createAsyncLocalStorage("deviceWallet"));
          break;
        case "credentialStore":
          wallet = await DeviceWalletImpl.fromCredentialStore();
          break;
        default:
          // default to local storage
          wallet = await DeviceWalletImpl.fromAsyncStorage(this.options.storage || createAsyncLocalStorage("deviceWallet"));
      }
      this.connector = new DeviceWalletConnector({
        chain: this.options.chain || Ethereum,
        wallet,
        chains: this.options.chains || defaultChains
      });
      _classPrivateFieldSet(this, _walletImpl, wallet);
    }
    return this.connector;
  }
  getWalletData() {
    if (!_classPrivateFieldGet(this, _walletImpl)) {
      throw new Error("Wallet not initialized");
    }
    return _classPrivateFieldGet(this, _walletImpl).getWalletData();
  }
  static getAddressStorageKey() {
    return STORAGE_KEY_ADDR;
  }
  static getDataStorageKey() {
    return STORAGE_KEY_DATA;
  }
}
_defineProperty(DeviceBrowserWallet, "id", "deviceWallet");
_defineProperty(DeviceBrowserWallet, "meta", {
  name: "Device Wallet",
  iconURL: "ipfs://QmcNddbYBuQKiBFnPcxYegjrX6S6z9K1vBNzbBBUJMn2ox/device-wallet.svg"
});
var _wallet = /*#__PURE__*/new WeakMap();
class DeviceWalletImpl extends AbstractWallet {
  static async fromAsyncStorage(storage) {
    return new DeviceWalletImpl({
      storage: new AsyncWalletStorage(storage)
    });
  }
  static async fromCredentialStore() {
    return new DeviceWalletImpl({
      storage: new CredentialsStorage(navigator.credentials)
    });
  }
  constructor(options) {
    super();
    _defineProperty(this, "options", void 0);
    _classPrivateFieldInitSpec(this, _wallet, {
      writable: true,
      value: void 0
    });
    this.options = options;
  }
  async getSigner(provider) {
    if (!_classPrivateFieldGet(this, _wallet)) {
      throw new Error("Wallet not initialized");
    }
    let wallet = _classPrivateFieldGet(this, _wallet);
    if (provider) {
      wallet = wallet.connect(provider);
    }
    return wallet;
  }
  async getSavedWalletAddress() {
    const data = await this.options.storage.getWalletData();
    if (!data) {
      return null;
    }
    return data.address;
  }
  async generateNewWallet() {
    const wallet = ethers.Wallet.createRandom();
    _classPrivateFieldSet(this, _wallet, wallet);
    return wallet.address;
  }
  async loadSavedWallet(password) {
    const data = await this.options.storage.getWalletData();
    if (!data) {
      throw new Error("No saved wallet");
    }
    const wallet = await ethers.Wallet.fromEncryptedJson(data.encryptedData, password);
    _classPrivateFieldSet(this, _wallet, wallet);
    return wallet.address;
  }
  async save(password) {
    const wallet = await this.getSigner();
    // reduce the scrypt cost to make it faster
    const options = {
      scrypt: {
        N: 1 << 32
      }
    };
    const encryptedData = await wallet.encrypt(password, options);
    await this.options.storage.storeWalletData({
      address: wallet.address,
      encryptedData
    });
  }
  async export(password) {
    const wallet = await this.getSigner();
    return wallet.encrypt(password);
  }
  getWalletData() {
    return this.options.storage.getWalletData();
  }
}
class AsyncWalletStorage {
  constructor(storage) {
    _defineProperty(this, "storage", void 0);
    this.storage = storage;
  }
  async getWalletData() {
    const [address, encryptedData] = await Promise.all([this.storage.getItem(STORAGE_KEY_ADDR), this.storage.getItem(STORAGE_KEY_DATA)]);
    if (!address || !encryptedData) {
      return null;
    }
    return {
      address,
      encryptedData
    };
  }
  async storeWalletData(data) {
    await Promise.all([this.storage.setItem(STORAGE_KEY_ADDR, data.address), this.storage.setItem(STORAGE_KEY_DATA, data.encryptedData)]);
  }
}
class CredentialsStorage {
  constructor(container) {
    _defineProperty(this, "container", void 0);
    this.container = container;
  }
  async getWalletData() {
    const credential = await this.container.get({
      password: true,
      unmediated: true
    });
    if (credential && "password" in credential) {
      return {
        address: credential.id,
        encryptedData: credential.password
      };
    }
    return null;
  }
  async storeWalletData(data) {
    if ("PasswordCredential" in window) {
      let credentialData = {
        id: data.address,
        password: data.encryptedData
      };
      const credential = await this.container.create({
        password: credentialData
      });
      if (!credential) {
        throw new Error("Credential not created");
      }
      await this.container.store(credential);
    } else {
      throw new Error("PasswordCredential not supported");
    }
  }
}

export { DeviceBrowserWallet, DeviceWalletImpl };
