// @thirdweb-dev/react/evm/dist/thirdweb-dev-react-evm.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var useSafe = require('../../dist/useSafe-4231d834.cjs.dev.js');
var useWalletConnect = require('../../dist/useWalletConnect-4a5fdd6c.cjs.dev.js');
var reactCore = require('@thirdweb-dev/react-core');
require('react');
require('react/jsx-runtime');
require('@tanstack/react-query');
require('@emotion/react');
require('@emotion/styled');
require('@radix-ui/react-popover');
require('copy-to-clipboard');
require('@radix-ui/react-tooltip');
require('@radix-ui/react-icons');
require('ethers');
require('@radix-ui/react-tabs');
require('@radix-ui/react-dropdown-menu');
require('@thirdweb-dev/chains');
require('tiny-invariant');
require('@radix-ui/colors');
require('@radix-ui/react-dialog');
require('detect-browser');
require('@thirdweb-dev/wallets');
require('react-qr-code');



exports.useSafe = useSafe.useSafe;
exports.ConnectWallet = useWalletConnect.ConnectWallet;
exports.MediaRenderer = useWalletConnect.MediaRenderer;
exports.ThirdwebNftMedia = useWalletConnect.ThirdwebNftMedia;
exports.ThirdwebProvider = useWalletConnect.ThirdwebProvider;
exports.Web3Button = useWalletConnect.Web3Button;
exports.useCoinbaseWallet = useWalletConnect.useCoinbaseWallet;
exports.useInstalledWallets = useWalletConnect.useInstalledWallets;
exports.useMetamask = useWalletConnect.useMetamask;
exports.usePaperWallet = useWalletConnect.usePaperWallet;
exports.usePaperWalletUserEmail = useWalletConnect.usePaperWalletUserEmail;
exports.useResolvedMediaType = useWalletConnect.useResolvedMediaType;
exports.useWalletConnect = useWalletConnect.useWalletConnect;
exports.useWalletConnectV1 = useWalletConnect.useWalletConnectV1;
Object.keys(reactCore).forEach(function (k) {
  if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return reactCore[k]; }
  });
});
